fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/ColumnEnforcer.java

```diff
@@ -0,0 +1,109 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+package com.starrocks.sql.optimizer.rule.transformation.materialization;
+
+import autovalue.shaded.com.google.common.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.starrocks.catalog.Column;
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.OptExpressionVisitor;
+import com.starrocks.sql.optimizer.operator.OperatorBuilderFactory;
+import com.starrocks.sql.optimizer.operator.Projection;
+import com.starrocks.sql.optimizer.operator.logical.LogicalOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalScanOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+
+import java.util.List;
+import java.util.Map;
+
+public class ColumnEnforcer {
+    private OptExpression optExpression;
+    private List<ColumnRefOperator> columnsToEnforce;
+    private EnforceContext enforceContext;
+
+    public ColumnEnforcer(
+            OptExpression optExpression, List<ColumnRefOperator> columnsToEnforce) {
+        this.optExpression = optExpression;
+        this.columnsToEnforce = columnsToEnforce;
+        this.enforceContext = new EnforceContext();
+    }
+
+    public OptExpression enforce() {
+        ColumnEnforcerVisitor visitor = new ColumnEnforcerVisitor();
+        OptExpression rootOptExpression = optExpression.getOp().accept(visitor, optExpression, enforceContext);
+        return rootOptExpression;
+    }
+
+    public List<ColumnRefOperator> getEnforcedColumns() {
+        return enforceContext.enforcedColumns;
+    }
+
+    private class EnforceContext {
+        public List<ColumnRefOperator> enforcedColumns;
+
+        public EnforceContext() {
+            this.enforcedColumns = Lists.newArrayList();
+        }
+    }
+
+    private class ColumnEnforcerVisitor extends OptExpressionVisitor<OptExpression, EnforceContext> {
+        @Override
+        public OptExpression visitLogicalTableScan(OptExpression optExpression, EnforceContext context) {
+            LogicalScanOperator scanOperator = optExpression.getOp().cast();
+            LogicalScanOperator.Builder builder = OperatorBuilderFactory.build(scanOperator);
+            builder.withOperator(scanOperator);
+            Map<ColumnRefOperator, Column> columnRefOperatorColumnMap = Maps.newHashMap(scanOperator.getColRefToColumnMetaMap());
+            for (ColumnRefOperator columnRef : columnsToEnforce) {
+                for (Map.Entry<Column, ColumnRefOperator> entry : scanOperator.getColumnMetaToColRefMap().entrySet()) {
+                    if (entry.getValue().equals(columnRef)) {
```

Comment:
how about use `scanOperator.getColRefToColumnMetaMap()` to check ?

Reply:
first, it has been checked in line 1420-1424.
secondly, if enforce failed, rewrite predicate will fail again, it will lead to rewrite failure